python_modules/dagster/dagster/_core/execution/run_metrics_thread.py

```diff
@@ -0,0 +1,296 @@
+        logging.debug(f"  [python_metrics_enabled={python_metrics_enabled}]")
+
+    while not shutdown_event.is_set():
+        try:
+            metrics = {}
+
+            if container_metrics_enabled:
+                container_metrics = _get_container_metrics(logger=logger)
+                metrics.update(container_metrics)
+
+            if python_metrics_enabled:
+                python_metrics = _get_python_runtime_metrics()
+                metrics.update(python_metrics)
+
+            if len(metrics) > 0:
+                _report_run_metrics_graphql(
+                    instance,
+                    dagster_run=dagster_run,
+                    metrics=metrics,
+                    run_tags=run_tags,
+                )
+
+            if container_metrics_enabled and report_container_metrics_as_engine_events:
+                _report_run_metrics_engine_event(instance, dagster_run, metrics)
+        except:
+            logging.error(
+                "Exception during capture of metrics, will cease capturing", exc_info=True
+            )
+            return False  # terminate the thread safely without interrupting the main thread
+
+        sleep(polling_interval)
+    if logger:
+        logging.debug("Shutting down metrics capture thread")
+    return True
+
+
+def start_run_metrics_thread(
+    instance: DagsterInstance,
+    dagster_run: DagsterRun,
+    container_metrics_enabled: Optional[bool] = True,
+    python_metrics_enabled: Optional[bool] = False,
+    logger: Optional[logging.Logger] = None,
+    polling_interval: float = DEFAULT_RUN_METRICS_POLL_INTERVAL_SECONDS,
+    report_container_metrics_as_engine_events: Optional[bool] = False,
+) -> Tuple[threading.Thread, threading.Event]:
+    check.inst_param(instance, "instance", DagsterInstance)
+    check.inst_param(dagster_run, "dagster_run", DagsterRun)
+    check.opt_inst_param(logger, "logger", logging.Logger)
+    check.opt_bool_param(container_metrics_enabled, "container_metrics_enabled")
+    check.opt_bool_param(python_metrics_enabled, "python_metrics_enabled")
+    check.float_param(polling_interval, "polling_interval")
+    check.opt_bool_param(
+        report_container_metrics_as_engine_events, "report_container_metrics_as_engine_events"
+    )
+
+    container_metrics_enabled = container_metrics_enabled and _process_is_containerized()
+
+    # TODO - ensure at least one metrics source is enabled
+    assert container_metrics_enabled or python_metrics_enabled, "No metrics enabled"
+
+    if logger:
+        logger.debug("Starting run metrics thread")
+
+    instance.report_engine_event(
+        f"Starting run metrics thread with container_metrics_enabled={container_metrics_enabled} and "
+        f"python_metrics_enabled={python_metrics_enabled}",
+        dagster_run=dagster_run,
+        run_id=dagster_run.run_id,
+    )
+
+    shutdown_event = threading.Event()
+    thread = threading.Thread(
```

Comment:
Adding threads in execution processes makes me a little nervous. Worth being extra careful with at least. For example we should make sure this is daemonic to prevent this thread from blocking process exit if it gets stuck. 

In the default multiprocess execution this will only be running with system run orchestration code. 

In todays in process executor or a future in-process-concurrent executor this thread will run along side user code.

Reply:
I think this is safe as is.

- the main thread doesn't depend or synchronize with that thread directly (however the thread does use RunStorage, which is based on a requests.Session - not fully thread safe but I don't think in a way that would cause deadlock).
- the termination of the main thread will timeout on join if the other thread is blocked.
- i am trapping exception and terminate that thread on errors because it's not essential.

Reply:
I noticed that the other similarish run cancellation thread is also not daemonic: https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_core/execution/run_cancellation_thread.py#L38-L42

Is there a reason not to make it daemonic here though? What's the downside?